internal/collectors/builder.go

```diff
@@ -0,0 +1,448 @@
+}
+
+func (b *Builder) buildResourceQuotaCollector() *coll.Collector {
+	filteredMetricFamilies := coll.FilterMetricFamilies(b.whiteBlackList, resourceQuotaMetricFamilies)
+	composedMetricGenFuncs := coll.ComposeMetricGenFuncs(filteredMetricFamilies)
+
+	familyHeaders := coll.ExtractMetricFamilyHeaders(filteredMetricFamilies)
+
+	store := metricsstore.NewMetricsStore(
+		familyHeaders,
+		composedMetricGenFuncs,
+	)
+	reflectorPerNamespace(b.ctx, b.kubeClient, &v1.ResourceQuota{}, store, b.namespaces, createResourceQuotaListWatch)
+
+	return coll.NewCollector(store)
+}
+
+func (b *Builder) buildSecretCollector() *coll.Collector {
+	filteredMetricFamilies := coll.FilterMetricFamilies(b.whiteBlackList, secretMetricFamilies)
+	composedMetricGenFuncs := coll.ComposeMetricGenFuncs(filteredMetricFamilies)
+
+	familyHeaders := coll.ExtractMetricFamilyHeaders(filteredMetricFamilies)
+
+	store := metricsstore.NewMetricsStore(
+		familyHeaders,
+		composedMetricGenFuncs,
+	)
+	reflectorPerNamespace(b.ctx, b.kubeClient, &v1.Secret{}, store, b.namespaces, createSecretListWatch)
+
+	return coll.NewCollector(store)
+}
+
+func (b *Builder) buildServiceCollector() *coll.Collector {
+	filteredMetricFamilies := coll.FilterMetricFamilies(b.whiteBlackList, serviceMetricFamilies)
+	composedMetricGenFuncs := coll.ComposeMetricGenFuncs(filteredMetricFamilies)
+
+	familyHeaders := coll.ExtractMetricFamilyHeaders(filteredMetricFamilies)
+
+	store := metricsstore.NewMetricsStore(
+		familyHeaders,
+		composedMetricGenFuncs,
+	)
+	reflectorPerNamespace(b.ctx, b.kubeClient, &v1.Service{}, store, b.namespaces, createServiceListWatch)
+
+	return coll.NewCollector(store)
+}
+
+func (b *Builder) buildStatefulSetCollector() *coll.Collector {
+	filteredMetricFamilies := coll.FilterMetricFamilies(b.whiteBlackList, statefulSetMetricFamilies)
+	composedMetricGenFuncs := coll.ComposeMetricGenFuncs(filteredMetricFamilies)
+
+	familyHeaders := coll.ExtractMetricFamilyHeaders(filteredMetricFamilies)
+
+	store := metricsstore.NewMetricsStore(
+		familyHeaders,
+		composedMetricGenFuncs,
+	)
+	reflectorPerNamespace(b.ctx, b.kubeClient, &apps.StatefulSet{}, store, b.namespaces, createStatefulSetListWatch)
+
+	return coll.NewCollector(store)
+}
+
+func (b *Builder) buildPodCollector() *coll.Collector {
+	filteredMetricFamilies := coll.FilterMetricFamilies(b.whiteBlackList, podMetricFamilies)
+	composedMetricGenFuncs := coll.ComposeMetricGenFuncs(filteredMetricFamilies)
+
+	familyHeaders := coll.ExtractMetricFamilyHeaders(filteredMetricFamilies)
+
+	store := metricsstore.NewMetricsStore(
+		familyHeaders,
+		composedMetricGenFuncs,
+	)
+	reflectorPerNamespace(b.ctx, b.kubeClient, &v1.Pod{}, store, b.namespaces, createPodListWatch)
+
+	return coll.NewCollector(store)
+}
+
+// reflectorPerNamespace creates a Kubernetes client-go reflector with the given
+// listWatchFunc for each given namespace and registers it with the given store.
+func reflectorPerNamespace(
```

Comment:
Can we make this a public func? 
Here is the code I copied to openshift-state-metrics https://github.com/wanghaoran1988/openshift-state-metrics/blob/ce6b3351e63e92f3eb8aa38edd926a65949baecb/pkg/collectors/builder.go#L257

Reply:
As a first thought I would have suggested to move it to the `metrics_store` package. Thinking about it a bit more, this would tightly couple the `metrics_store` to the specifics of client_golang and the vanilla `kubeClient`.

Given that this is a small function, I would again ["prefer duplication over the wrong abstraction"](https://www.sandimetz.com/blog/2016/1/20/the-wrong-abstraction). Not a strong opinion. I am curious what you think.

Reply:
I like the "prefer duplication over the wrong abstractions" blog

Reply:
Then moving this to `pkg/collector` and making it public sgty? Would that still be useful @wanghaoran1988 ?

Reply:
@LiliC No, you don't need make it public now, due to mine need different params, I'd like have my copy there.

Reply:
Yes same for me, I have a custom reflector 👍